src/desktop/widgets/rulerwidget.cpp

```diff
@@ -0,0 +1,162 @@
+		qreal dist = std::abs(initialPPD - ppd);
+		if(dist < bestDist) {
+			adjustedPPD = ppd;
+			bestDist = dist;
+		}
+	}
+
+	int pixelsPerRulerDivision =
+		std::lround((adjustedPPD * m_canvasToRulerScale));
+	divisions = longSize() / pixelsPerRulerDivision;
+
+	QColor dark = palette().color(QPalette::Dark);
+	QColor text = palette().color(QPalette::Text);
+
+	painter.setPen(dark);
+	painter.setRenderHint(QPainter::Antialiasing, true);
+	// Set pen for consistent line width on high DPI displays
+	QPen pen = painter.pen();
+	pen.setWidthF(1.0 / devicePixelRatioF());
+	painter.setPen(pen);
+
+	qreal startOffset = std::fmod(std::abs(m_lMin), adjustedPPD);
+	int canvasL = std::lround(
+		m_lMin < 0 ? (m_lMin + startOffset) - adjustedPPD
+				   : m_lMin - startOffset);
+
+	int lineHeight = 0;
+	if(!isHorizontal()) {
+		// Figure out the largest number we're going to draw so that we can
+		// choose a smaller font size if it's too big.
+		auto digitsNeeded = [&](int num) {
+			int negative = 0;
+			if(num < 0) {
+				num = -num;
+				negative = 1;
+			}
+			for(int i = 1; i <= std::numeric_limits<int>::digits10; ++i) {
+				if(num < std::pow(10, i))
+					return i + negative;
+			}
+			return std::numeric_limits<int>::digits10 + 1 + negative;
+		};
+
+		int digits = std::max(
+			digitsNeeded(canvasL),
+			digitsNeeded(canvasL + adjustedPPD * divisions));
+
+		while(true) {
+			lineHeight =
+				painter.fontMetrics().tightBoundingRect("0123456789").height();
+			lineHeight += 2;
+
+			if(lineHeight * digits >= pixelsPerRulerDivision - 2) {
+				QFont smallerFont = painter.font();
+				int newPointSize = smallerFont.pointSize() - 1;
+				if(newPointSize == 0)
+					// The font doesn't seem to be responding to point size or
+					// something else is wrong. Bail out here and continue with
+					// the current font.
+					break;
+				smallerFont.setPointSize(newPointSize);
+				painter.setFont(smallerFont);
+			} else
+				break;
+		}
+	}
+
+	for(;; canvasL += adjustedPPD) {
+		int l = canvasToRuler(canvasL);
+		if(l < -pixelsPerRulerDivision)
+			continue;
+		if(l >= longSize())
+			break;
+		painter.setPen(dark);
+		painter.drawLine(adjustPoint(l, 0), adjustPoint(l, shortSize()));
+		painter.setPen(text);
+		if(isHorizontal())
+			painter.drawText(
+				adjustPoint(l + 2, shortSize() - 2), QString::number(canvasL));
+		else {
+			// Qt can't draw anti-aliased texted when rotated, so draw it
+			// vertically. Also, we can't set the line spacing using drawText,
+			// so we get to draw each character manually.
+			QString label = QString::number(canvasL);
+			int offset = lineHeight;
+			for(QChar c : label) {
+				painter.drawText(2, l + offset, c);
+				offset += lineHeight;
+			}
+		}
+	}
+}
+
+void RulerWidget::onViewChanged(const QPolygonF &view)
+{
+	if(isHorizontal()) {
+		m_lMin = view.boundingRect().left();
+		m_lMax = view.boundingRect().right();
+	} else {
+		m_lMin = view.boundingRect().top();
+		m_lMax = view.boundingRect().bottom();
+	}
```

Comment:
The widget sometimes doesn't change for me when the canvas is moved because it doesn't schedule repaints. Functions that change painting-related state need to call `QWidget::update` to let Qt know that it's painting time.

```suggestion
	}
	update();
```